src/Extension/FormSpamProtectionExtension.php

```diff
@@ -109,7 +110,8 @@ public function enableSpamProtection($options = array())
         $protector = self::get_protector($options);
 
         if ($protector === null) {
-            throw new LogicException('No spam protector has been set. Null is not valid value.');
+            Requirements::customScript('console.error("No spam protector has been set on this form.")');
+            return $this->owner;
```

Comment:
I don't think we should change this. As mentioned in https://github.com/silverstripe/silverstripe-spamprotection/issues/103, when we upgraded this module for CMS 5 compatibility and dropped support for the [silverstripe/akismet](https://packagist.org/packages/silverstripe/akismet) module we made a conscious decision to leave this exception here - developers should be made aware early on that there's a missing dependency, and may miss a console error.

Reply:
@GuySartorelli this is for an environment where no recaptcha key is set, that should not kill the entire web page.

Reply:
This module doesn't know anything about recaptcha - a separate module would need to be used (e.g. [undefinedoffset/silverstripe-nocaptcha](https://github.com/UndefinedOffset/silverstripe-nocaptcha)) and set as the spam protector following [these instructions in the README](https://github.com/silverstripe/silverstripe-spamprotection?tab=readme-ov-file#configuring).

That separate module is then in charge of how it handles missing configuration e.g. missing recaptcha keys. If you're calling `enableSpamProtection()`, then there absolutely should be a protector set - otherwise there has been a developer error. I can't see any way this _shouldn't_ throw an error for the developer to resolve.

Reply:
@GuySartorelli I disagree completely (website going down and showing an error to users is absolutely last resort) but I've put my decision behind a config flag for now so it can be toggled off at least.

Reply:
Sorry, I'm just struggling to understand how you'd get into a situation where this happens on a production website without a developer having encountered the exception first and remedied the configuration... can you please help me understand the scenario in which what would happen?

Reply:
@GuySartorelli there could be tons of reasons why this happens we can't predict everything. The scenario at least for me which triggered this PR is ops team migrated server instances, missed the recaptcha key from the pipeline and took the entire site down (since the form was in the footer across the site). 

But I can imagine other cases i.e on localhost or spam protection being disabled on CI / testing where you don't want the spam protection to be required (and you don't want to change the code)

Reply:
You've mentioned recaptcha keys before... but again, this module doesn't know anything about recaptcha. This exception is triggered when `SilverStripe\SpamProtection\Extension\FormSpamProtectionExtension.default_spam_protector` hasn't been set, which is a yaml configuration setting that should be checked into version control.

Can you please help me understand how a missing recaptcha key (which is most likely associated with another module which implements a spam protector got google recaptcha, rather than directly relating to this module) resulted in this exception triggering?

Reply:
In our case `default_spam_protector` had an `envarset` wrapper. So without this environment variable set there was no `default_spam_protector` provided.

Reply:
In that case, it sounds like you need to either set yourself a default null spam protector in the case that environment variable _isn't_ set, or revisit why that environment variable would determine the setting of a _default_ spam protector in the first place.

I don't think this module should throw console errors (which will go unnoticed) to indicate a configuration failure like this. Even with that scenario in mind, which seems to me like an incorrect way to configure the default spam protector, I _still_ think it should throw an exception.

For your specific case, you could either set a null default spam protector when your environment variable dependency is not set, or add try/catch around your calls to `enableSpamProtection()`, or override the extension with a custom implementation that allows for not having a spam protector set, etc - but I really do think this module needs to explicitly throw an exception if there is no spam protector when you call `enableSpamProtection()`.

Any thoughts from the rest of @silverstripe/core-team?

Reply:
If this is for use in tandem with https://github.com/undefinedoffset/silverstripe-nocaptcha, you can also set default values with fake recaptcha keys in yaml when your environment variable isn't set - which will mean the recaptcha doesn't work (because the key is invalid), but the page still loads.

Reply:
@GuySartorelli it doesn't throw console errors - it uses a PHP warning now. You're looking at an outdated commit

Reply:
> Any thoughts from the rest of https://github.com/orgs/silverstripe/teams/core-team?

Throw hard exceptions if it's something is misconfigured so people get immediate feedback is the general philosophy we follow.  Silent errors are generally bad, means things can be broken for a long time before anyone realises.

@wilr Sorry to hear about you about woes with your migration, though sounds like what happened there is a lack of testing as part of the migration.  I don't see that as a good reason to change our general philosophy

Reply:
Ok given the change as now does not change behaviour I don't see any problem merging?

Reply:
The problem as I see it is that this just doesn't seem like a good change. I don't think we should make it easier to ignore misconfigured environments - and there are already workarounds if you _do_ want to, for some reason.

Reply:
I don't believe any of the work arounds are very user friendly. My original comment is still, having no spam protection is better than a broken website.